sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingLogFileSystem.scala

```diff
@@ -353,6 +353,233 @@ private[sharing] object DeltaSharingLogFileSystem extends Logging {
+   */
+  def constructLocalDeltaLogAcrossVersions(
+      lines: Seq[String],
+      customTablePath: String,
+      startingVersionOpt: Option[Long],
+      endingVersionOpt: Option[Long]): ConstructedDeltaLogMetadata = {
+    val startTime = System.currentTimeMillis()
+    assert(
+      startingVersionOpt.isDefined == endingVersionOpt.isDefined,
+      s"startingVersionOpt($startingVersionOpt) and endingVersionOpt($endingVersionOpt) should be" +
+      " both defined or not."
+    )
+    if (startingVersionOpt.isDefined) {
+      assert(
+        startingVersionOpt.get <= endingVersionOpt.get,
+        s"startingVersionOpt($startingVersionOpt) must be smaller than " +
+        s"endingVersionOpt($endingVersionOpt)."
+      )
+    }
+    var minVersion = Long.MaxValue
+    var maxVersion = 0L
+    var minUrlExpirationTimestamp: Option[Long] = None
+    val idToUrl = scala.collection.mutable.Map[String, String]()
+    val versionToDeltaSharingFileActions =
+      scala.collection.mutable.Map[Long, ArrayBuffer[model.DeltaSharingFileAction]]()
+    val versionToMetadata = scala.collection.mutable.Map[Long, model.DeltaSharingMetadata]()
+    val versionToJsonLogBuilderMap = scala.collection.mutable.Map[Long, ArrayBuffer[String]]()
+    val versionToJsonLogSize = scala.collection.mutable.Map[Long, Long]().withDefaultValue(0L)
+    var numFileActionsInMinVersion = 0
+    val versionToTimestampMap = scala.collection.mutable.Map[Long, Long]()
+    var startingMetadataLineOpt: Option[String] = None
+    var startingProtocolLineOpt: Option[String] = None
+
+    lines.foreach { line =>
+      val action = JsonUtils.fromJson[model.DeltaSharingSingleAction](line).unwrap
+      action match {
+        case fileAction: model.DeltaSharingFileAction =>
+          minVersion = minVersion.min(fileAction.version)
+          maxVersion = maxVersion.max(fileAction.version)
+          // Store file actions in an array to sort them based on id later.
+          versionToDeltaSharingFileActions.getOrElseUpdate(
+            fileAction.version,
+            ArrayBuffer[model.DeltaSharingFileAction]()
+          ) += fileAction
+        case metadata: model.DeltaSharingMetadata =>
+          if (metadata.version != null) {
+            // This is to handle the cdf and streaming query result.
+            minVersion = minVersion.min(metadata.version)
+            maxVersion = maxVersion.max(metadata.version)
+            versionToMetadata(metadata.version) = metadata
+            if (metadata.version == minVersion) {
+              startingMetadataLineOpt = Some(metadata.deltaMetadata.json + "\n")
+            }
+          } else {
+            // This is to handle the snapshot query result from DeltaSharingSource.
+            startingMetadataLineOpt = Some(metadata.deltaMetadata.json + "\n")
+          }
+        case protocol: model.DeltaSharingProtocol =>
+          startingProtocolLineOpt = Some(protocol.deltaProtocol.json + "\n")
```

Comment:
do we expect to get only one protocol action from the server? what if we have multiple protocol actions (due to alter etc) in the log files between starting and end version?

Reply:
Same qn for metadata.

Reply:
good question, the answer is two fold:
1) The server performs basic checks that the recipient/client is able to handle the readerFeatures, and throw error if not. 
2) The rest of the job is left to the client side delta library, and fallback to the same delta behavior. if delta cannot handle it, it throws the same error as reading a delta table, such as error when reading cdf on a table with column name changed.